R/nhgis_read.r

```diff
@@ -3,255 +3,666 @@
-  join_vars <- fostr_subset(join_vars, "GISJOIN.*")
+  cb_files <- find_files_in(
+    data_file,
+    name_ext = "txt",
+    multiple_ok = TRUE,
+    none_ok = TRUE
+  )
+
+  if (length(cb_files) > 0) {
+    cb_file <- fostr_subset(
+      cb_files,
+      fostr_replace(
+        basename(file),
+        ipums_file_ext(file),
+        ""
+      )
+    )
+  }
 
-  # Drop overlapping vars besides join var from shape file
-  drop_vars <- dplyr::intersect(names(data), names(sf_data))
-  drop_vars <- dplyr::setdiff(drop_vars, join_vars)
-  sf_data <- dplyr::select(sf_data, -one_of(drop_vars))
+  cb_ddi_info <- try(
+    read_nhgis_codebook(data_file, file_select = tidyselect::all_of(cb_file)),
+    silent = TRUE
+  )
 
-  # Avoid a warning by adding attributes from the join_vars in data to
-  # join_vars in sf_data
-  purrr::walk(join_vars, function(vvv) {
-    attributes(sf_data[[vvv]]) <<- attributes(data[[vvv]])
-  })
+  cb_error <- inherits(cb_ddi_info, "try-error")
 
-  # Coerce to data.frame to avoid sf#414 (fixed in development version of sf)
-  data <- dplyr::full_join(as.data.frame(sf_data), as.data.frame(data), by = join_vars)
-  data <- sf::st_as_sf(tibble::as_tibble(data))
+  if (cb_error) {
+    cb_ddi_info <- NHGIS_EMPTY_DDI
+  }
 
-  # Check if any data rows are missing (merge failures where not in shape file)
   if (verbose) {
-    missing_in_shape <- purrr::map_lgl(data$geometry, is.null)
-    if (any(missing_in_shape)) {
-      gis_join_failures <- data$GISJOIN[missing_in_shape]
-      cat(paste(
-        custom_format_text(
-          "There are ", sum(missing_in_shape), " rows of data that ",
-          "have data but no geography. This can happen because:"
-        ),
-        custom_format_text(
-          "Shape files do not include some census geographies such ",
-          "as 'Crews of Vessels' tracts that do not have a defined area",
-          indent = 2, exdent = 2
-        ),
-        custom_format_text(
-          "Shape files have been simplified which sometimes drops ",
-          "entire geographies (especially small ones)."
-        ),
-        sep = "\n"
-      ))
-    }
+    message(short_conditions_text(cb_ddi_info))
+  }
+
+  if (file_is_zip(data_file)) {
+    file <- unz(data_file, file)
+  } else if (file_is_dir(data_file)) {
+    file <- file.path(data_file, file)
   }
+
+  header_info <- check_header_row(data_file, file_select = !!file_select)
+
+  if (header_info$has_extra_header && remove_extra_header) {
+    skip <- 2 + skip
+  } else {
+    skip <- 1 + skip
+  }
+
+  if (is_null(col_names) || isTRUE(col_names)) {
+    col_names <- header_info$col_names
+  }
+
+  # Specify encoding (assuming all nhgis extracts are ISO-8859-1 eg latin1
+  # because an extract with county names has n with tildes and so is can
+  # be verified as ISO-8859-1)
+  cb_ddi_info$file_encoding <- "ISO-8859-1"
+  locale <- locale %||% ipums_locale(cb_ddi_info$file_encoding)
+
+  data <- readr::read_csv(
```

Comment:
Discussion topic: Not sure if this would be possible, but if it is, what would the pros and cons be of using the codebook to specify the col_types to `read_csv()`? I don't know how relevant this is for NHGIS data, but I have gotten into the habit of explicitly specifying the col_types when I use `read_csv()`, so it is kind of jarring to see the col_types specification printed out when I call `read_nhgis()`.

Reply:
NHGIS .txt codebooks don't provide info on col types so I don't think it's doable from the codebook. But since it now allows argument passing to readr users can now specify either `col_types` or `show_col_types` to specify or suppress. We could consider pulling `col_types` out of the dots to make it more clear that it's an option though.